Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp

```diff
@@ -1766,6 +1766,11 @@ int usartFifoAvailable(SZ_USART_FIFO *fifo) {
   return WIFISERIAL.available();
 }
 
+constexpr bool str_match(const char * const string, const char * const pattern, const int index=0) {
+  const char c = pattern[index];
+  return (string[index] == c) && (c == '\0' || str_match(string, pattern, index + 1));
+}
+
```

Comment:
I'm almost sure that what the compiler is already doing something like this for plain old strcmp.
It's not optimizing like you expect (see [here](https://godbolt.org/z/KPKPWz), it's not converting to a chain of comparison).

Reply:
The code there only has one `strcmp` instead of three with `else`. Let's see what it does with that…. And also see what `strcmp_P` looks like….

Reply:
Here is a clear demo of how much better the compiler's `strcmp` is compared to `str_match`:
```cpp
#include <stdio.h>
#include <stdlib.h>
#include <climits>
#include <cstring>

constexpr bool str_match(const char * const string, const char * const pattern) {
  return *string == *pattern && (*pattern == '\0' || str_match(string + 1, pattern + 1));
}

static bool check_string(const char * const txt) {
         if (str_match(txt, "M108")) { printf("match M108\n"); return true; }
    else if (str_match(txt, "M112")) { printf("match M112\n"); return true; }
    else if (str_match(txt, "M410")) { printf("match M410\n"); return true; }
    return false;
}

static bool check_string_cmp(const char * const txt) {
         if (strcmp(txt, "M108") == 0) { printf("match M108\n"); return true; }
    else if (strcmp(txt, "M112") == 0) { printf("match M112\n"); return true; }
    else if (strcmp(txt, "M410") == 0) { printf("match M410\n"); return true; }
    return false;
}

int main(int argc, char ** argv) {

    if (check_string(argv[0])) printf("String checks out with str_match.\n");

    if (check_string_cmp(argv[0])) printf("String checks out with strcmp.\n");

    return 0;
}
```

Reply:
The strcmp is a lot more compact. This is mainly due to the fact that it's not inlined compared to str_match, so in your example, you've 3 str_match instance for 1 strcmp instance.

disassembly:
```
_main:
0000000100000df7	pushq	%rbp
0000000100000df8	movq	%rsp, %rbp
0000000100000dfb	pushq	%r14
0000000100000dfd	pushq	%rbx
0000000100000dfe	movq	%rsi, %r14
0000000100000e01	movq	(%r14), %rbx
0000000100000e04	movl	$0x1, %ecx
0000000100000e09	movzbl	(%rbx), %eax
0000000100000e0c	cmpl	$0x4d, %eax
0000000100000e0f	jne	0x100000ea6
0000000100000e15	movb	$0x4d, %sil
0000000100000e18	leaq	0x121(%rip), %rdx ## literal pool for: "M108"
0000000100000e1f	movl	$0x1, %eax
0000000100000e24	testb	%sil, %sil
0000000100000e27	je	0x100000e79
0000000100000e29	movzbl	(%rbx,%rcx), %esi
0000000100000e2d	movzbl	(%rcx,%rdx), %edi
0000000100000e31	incq	%rcx
0000000100000e34	cmpl	%edi, %esi
0000000100000e36	je	0x100000e24
0000000100000e38	movb	$0x4d, %sil
0000000100000e3b	leaq	0x103(%rip), %rdx ## literal pool for: "M112"
0000000100000e42	movl	$0x1, %ecx
0000000100000e47	testb	%sil, %sil
0000000100000e4a	je	0x100000e82
0000000100000e4c	movzbl	(%rbx,%rax), %esi
0000000100000e50	movzbl	(%rax,%rdx), %edi
0000000100000e54	incq	%rax
0000000100000e57	cmpl	%edi, %esi
0000000100000e59	je	0x100000e47
0000000100000e5b	movb	$0x4d, %dl
0000000100000e5d	leaq	0xe6(%rip), %rax ## literal pool for: "M410"
0000000100000e64	testb	%dl, %dl
0000000100000e66	je	0x100000e8b
0000000100000e68	movzbl	(%rbx,%rcx), %edx
0000000100000e6c	movzbl	(%rcx,%rax), %esi
0000000100000e70	incq	%rcx
0000000100000e73	cmpl	%esi, %edx
0000000100000e75	je	0x100000e64
0000000100000e77	jmp	0x100000ea6
0000000100000e79	leaq	0x128(%rip), %rdi ## literal pool for: "match M108"
0000000100000e80	jmp	0x100000e92
0000000100000e82	leaq	0x114(%rip), %rdi ## literal pool for: "match M112"
0000000100000e89	jmp	0x100000e92
0000000100000e8b	leaq	0x100(%rip), %rdi ## literal pool for: "match M410"
0000000100000e92	callq	0x100000f10 ## symbol stub for: _puts
0000000100000e97	leaq	0xd2(%rip), %rdi ## literal pool for: "String checks out with str_match."
0000000100000e9e	callq	0x100000f10 ## symbol stub for: _puts
0000000100000ea3	movq	(%r14), %rbx
0000000100000ea6	leaq	0x93(%rip), %rsi ## literal pool for: "M108"
0000000100000ead	movq	%rbx, %rdi
0000000100000eb0	callq	0x100000f16 ## symbol stub for: _strcmp
0000000100000eb5	testl	%eax, %eax
0000000100000eb7	je	0x100000ee8
0000000100000eb9	leaq	0x85(%rip), %rsi ## literal pool for: "M112"
0000000100000ec0	movq	%rbx, %rdi
0000000100000ec3	callq	0x100000f16 ## symbol stub for: _strcmp
0000000100000ec8	testl	%eax, %eax
0000000100000eca	je	0x100000ef1
0000000100000ecc	leaq	0x77(%rip), %rsi ## literal pool for: "M410"
0000000100000ed3	movq	%rbx, %rdi
0000000100000ed6	callq	0x100000f16 ## symbol stub for: _strcmp
0000000100000edb	testl	%eax, %eax
0000000100000edd	jne	0x100000f09
0000000100000edf	leaq	0xac(%rip), %rdi ## literal pool for: "match M410"
0000000100000ee6	jmp	0x100000ef8
0000000100000ee8	leaq	0xb9(%rip), %rdi ## literal pool for: "match M108"
0000000100000eef	jmp	0x100000ef8
0000000100000ef1	leaq	0xa5(%rip), %rdi ## literal pool for: "match M112"
0000000100000ef8	callq	0x100000f10 ## symbol stub for: _puts
0000000100000efd	leaq	0x4c(%rip), %rdi ## literal pool for: "String checks out with strcmp."
0000000100000f04	callq	0x100000f10 ## symbol stub for: _puts
0000000100000f09	xorl	%eax, %eax
0000000100000f0b	popq	%rbx
0000000100000f0c	popq	%r14
0000000100000f0e	popq	%rbp
0000000100000f0f	retq
```